inc/connector/namespace.php

```diff
@@ -382,16 +382,39 @@ function pre_get_scheduled_event( $pre, $hook, $args, $timestamp ) {
 	}
 
 	$jobs = Job::get_jobs_by_query( [
+		'args' => null,
+		'hook' => null,
+		'limit' => 100,
+	] );
+
+	$filter = [
 		'hook' => $hook,
 		'timestamp' => $timestamp,
 		'args' => $args,
-	] );
+	];
+
+	// Filter results array.
+	$jobs = array_filter( $jobs, function ( $event ) use ( $filter ) : bool {
+		if ( is_string( $filter['hook'] ) && $event->hook !== $filter['hook'] ) {
+			return false;
+		}
+
+		if ( ! is_null( $filter['args'] ) && $event->args !== $filter['args'] ) {
+			return false;
+		}
+
+		if ( is_int( $filter['timestamp'] ) ) {
+			return $event->nextrun === $filter['timestamp'];
+		}
+
+		return true;
+	} );
 
 	if ( is_wp_error( $jobs ) || empty( $jobs ) ) {
 		return false;
 	}
 
-	$job = $jobs[0];
+	$job = array_shift( $jobs );
```

Comment:
[Array shift is very slow](https://cmljnelson.blog/2018/10/16/phps-array_shift-performance/) better to use array_pop

Reply:
That's pretty surprising tbh as it's kind of the same. Needs to be the first item in the array though. I used this because the array indexes aren't necessarily zero based after filtering.

Reply:
`array_shift()` is not the same operation; `array_shift()` modifies the array by removing the first element and then reindexing the array (`O(N)`), while `$job = $jobs[0]` just creates a new variable (`O(1)`). `array_pop()` modifies the array by removing the _last_ element, and hence does not need to reindex the array (`O(1)`), which is why it's faster.

@roborourke The most performant way would be `reset( $jobs )` which will reset the internal array iterator, then pull the first item, but better is if you can avoid this by having a zero-indexed array in the first place :D

Reply:
I’ve updated it to avoid array_shift